Cql pkg config and cql static lib

razvan-becheriu opened the pull request.

fxdupont commented:
When you think it is ready please put Kea 5488 under review.

razvan-becheriu commented:
Hi,

Sorry for the extra commits.
I was trying to make it possible to use the 'pkg-config' and the already
provided 'tools/cql_config' seamlessly.
I have tested both configuration scenarios and everything works OK now.

Just as I have mentioned in the ticket comments, it is a little bit
complicated to statically link with the cassandra_static library when using
pre-installed packages - the official rpm/deb package is not compiled with
the -fPIC flag and the compilation fails at link time.
To work around this error, you have to build the rpm/deb package yourself,
after you modify the default compilation flags in the cmake scripts in the
cpp driver directory.

___________________________________________________________________________________

Note:
to compile with static cpp-driver, the driver must be compiled with:
cmake .. -DCASS_BUILD_STATIC=ON -DCMAKE_CXX_FLAGS="-fPIC"

If you want to build the packages to use with apt or yum, please add
"-DCMAKE_CXX_FLAGS='-fPIC'" to the cmake command by modifying the lines in:

packaging/debian/rules
dh_auto_configure -- -DCMAKE_BUILD_TYPE=RELEASE -DCASS_BUILD_STATIC=ON
-DCMAKE_CXX_FLAGS='-fPIC' -DCASS_INSTALL_PKG_CONFIG=OFF
-DCMAKE_INSTALL_LIBDIR=/usr/lib

packaging/homebrew/cassandra-cpp-driver.rb
system "cmake", "-DCMAKE_BUILD_TYPE=RELEASE", "-DCASS_BUILD_STATIC=ON",
"-DCMAKE_CXX_FLAGS='-fPIC'", "-DCASS_INSTALL_PKG_CONFIG=OFF",
"-DCMAKE_INSTALL_PREFIX:PATH=#{prefix}", "-DCMAKE_INSTALL_LIBDIR=#{lib}",
".."
___________________________________________________________________________________

Regards,
Razvan


On Tue, Mar 6, 2018 at 2:41 PM, fxdupont <notifications@github.com> wrote:

> When you think it is ready please put Kea 5488 under review.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/isc-projects/kea/pull/71#issuecomment-370769187>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ASkW2wqQFxovknrWcDvujfKO85RZjt0Bks5tboP6gaJpZM4ScetQ>
> .
>



-- 
razvan becheriu
[image: qualitance.com]
senior software engineer
+40 (746) 090 415 <+40+(746)+090+415>, qualitance.com
<https://www.qualitance.com/>

razvan-becheriu commented:
This PR is ready for review.
It contains fixes for tickets 5488 and 5490.

Thanks,
Razvan

fxdupont commented:
Many comments:
 - first there are a lot of unrelated changes. Please keep only cql changes moving space/tab inconsistencies to another PR and for sure benchmark changes to their own ticket.

 - I disagree about the cql-static-link name: it is not static linking but linking with the static library, so simply replace link by library. BTW technically you should not build a dynamic object from a static library, at best you get a warning.

 - I can't see a reference to pkg-config and of course giving the pkg-config command to configure does not work. I tried with a cql_config script set to "pkg-config cpp-driver $*" (cassandra package name works too) on my macOS and built a working Kea. So there is something missing but it should be easy to fix... 

 - AC_DEFINE_SUBST does not exist. I believe you mean AC_DEFINE.

 - cql_config change for --libs is very incorrect: _static in the library name, system specific -lpthread

 - a remark: cpp-driver sources have 2 .pc files in packaging: one is for the static variant.

razvan-becheriu commented:
Hi,Regarding the spaces and google benchmarks, these changes are already included in PR 70 which I thought it will be merged first. Having this in mind, these differences should not appear in the diff.I will try to filter out these changes.Regarding static linking, I added this option because there is no official ubuntu deb package for the cpp driver.Datastax provides one, but it has not been included in the official ubuntu repository.If the kea libs link statically with the cpp driver which can be present on the build machine only, then, after install‎, no extra dependences are required on the target machine.I find no issue with the fact that the dynamic libs are statically linked to other libraries, it means that all the symbols needed will also be included in the dynamic lib, nothing more of course (not all the symbols in the cpp driver will be included, only the referenced ones)‎.I will change the name of the configure option to cql-static-lib.On ubuntu pkg-config is a tool that is installed on the system by default (I think). It is not provided by the xxxxx-lib-dev. If I understand correctly, there is no pkg-config executable in your ‎PATH. Maybe you can install it with brew.I am not sure what component you wish to install pkg-config. Please give more details.Regarding AC_DEFINE_SUBST, you are right. My bad. I will fix this.Regarding cql_config --libs, you are right. I have missed this. The dependency to cassandra_static‎ lib should be enabled only if USE_CQL_STATIC_LINK‎ (which I will change to USE_CQL_STATIC_LIB) is defined.Regarding -lphread. On Linux, this works fine. However I know that there is also the -pthread variant, but I am not sure which is more portable (on MAC OSX I believe that the -pthread is preferred). Please comment on this.‎I know about the two .pc variants.Thanks for the comments.I will try to get a Mac and test the build process myself.Regards,Razvan                                                                                                                                                                                                                                                                                                                                        Sent from my BlackBerry 10 smartphone on the Orange network.                                                                                                                                                                                                                From: fxdupontSent: Saturday, March 10, 2018 00:52To: isc-projects/keaReply To: isc-projects/keaCc: razvan-becheriu; AuthorSubject: Re: [isc-projects/kea] Cql pkg config and cql static link (#71)Many comments:


first there are a lot of unrelated changes. Please keep only cql changes moving space/tab inconsistencies to another PR and for sure benchmark changes to their own ticket.


I disagree about the cql-static-link name: it is not static linking but linking with the static library, so simply replace link by library. BTW technically you should not build a dynamic object from a static library, at best you get a warning.


I can't see a reference to pkg-config and of course giving the pkg-config command to configure does not work. I tried with a cql_config script set to "pkg-config cpp-driver $*" (cassandra package name works too) on my macOS and built a working Kea. So there is something missing but it should be easy to fix...


AC_DEFINE_SUBST does not exist. I believe you mean AC_DEFINE.


cql_config change for --libs is very incorrect: _static in the library name, system specific -lpthread


a remark: cpp-driver sources have 2 .pc files in packaging: one is for the static variant.



—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/isc-projects/kea","title":"isc-projects/kea","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/isc-projects/kea"}},"updates":{"snippets":[{"icon":"PERSON","message":"@fxdupont in #71: Many comments:\r\n - first there are a lot of unrelated changes. Please keep only cql changes moving space/tab inconsistencies to another PR and for sure benchmark changes to their own ticket.\r\n\r\n - I disagree about the cql-static-link name: it is not static linking but linking with the static library, so simply replace link by library. BTW technically you should not build a dynamic object from a static library, at best you get a warning.\r\n\r\n - I can't see a reference to pkg-config and of course giving the pkg-config command to configure does not work. I tried with a cql_config script set to \"pkg-config cpp-driver $*\" (cassandra package name works too) on my macOS and built a working Kea. So there is something missing but it should be easy to fix... \r\n\r\n - AC_DEFINE_SUBST does not exist. I believe you mean AC_DEFINE.\r\n\r\n - cql_config change for --libs is very incorrect: _static in the library name, system specific -lpthread\r\n\r\n - a remark: cpp-driver sources have 2 .pc files in packaging: one is for the static variant."}],"action":{"name":"View Pull Request","url":"https://github.com/isc-projects/kea/pull/71#issuecomment-371968253"}}}

razvan-becheriu commented:
You are right,
It seems that I have missed the line with the pkg-config. Fixed it now.

I was thinking that I should leave the spaces and the benchmarks in, and you should review this PR after PR70 is merged.
If I remove the spaces now, and you merge this PR after PR70, then all the changes will be undone.
Is it of to review this branch after PR 70 is merged?

Regards,
Razvan

fxdupont commented:
yes I use macOS (last official name, before it was OS X) with Xcode (the version with the SDK for my macOS version, less easy than it should be...), Xcode line commands (without them for instance after an Xcode upgrade you don't have /usr/include) and homebrew (an equivalent should work but brew is dedicated to macOS so easier, common and well supported). BTW they are needed to build Kea, and brew is needed to run Kea too.

About static stuff: some systems give warnings like this:
`*** Warning: Linking the shared library libutil_unittests.la against the static library ../../../../ext/gtest/libgtest.a is not portable!`
and worse I received a few days ago a message about something taking more time than scheduled because a static library in a package was compiled without `-fPIC`...
Now people using the static cassandra library are supposed to know what they do.

About -lpthread Kea uses threading so the required flags and libs are already linked: in conclusion simply remove it.

To come back to the main problem about pkg-config (and as I understood the subject of this PR) there are 3 points:
 - what is the library argument to use? According to cpp-driver sources it should be cassandra (or cassandra_static) from the .pc file name. In my brew formula I link it to cpp-driver but we should use a system independent name.
 - how to make cql_config "pkg-config compatible"? It already uses the same `--` arguments so the last step is to check if the first argument is the library name (cf first point) and to shift it.
 - how to support pkg-config? As soon as pkg-config and cql_config can be called with the same arguments it is enough to update messages to say pkg-config or cql_config. To be more user-friendly I propose to default to pkg-config got the `--with-cql` making the PATH optional (this is based on the remark that when pkg-config is available it is in the shell path).

razvan-becheriu commented:
Hi,

I have updated the cql_config script to accept only cassandra or cassandra_static parameters.
You can also run --help to view the command format and options.

On my Ubuntu, pkg-config can be called either with pkg-config modname --libs or pkg-config --libs modname. I have called (in configure.ac) pks-config with the second syntax (pkg-config --libs modname).
I have implemented this version in cql_config script (see --help).

The default --with-cql with no parameters calls pkg-config for the cassandra (dynamic) library.
You can try it on MacOSX.

Regards,
Razvan

fxdupont commented:
There is nothing called MacOSX, it was MacOS, be replaced by something very different named OS X which was recently renamed macOS... I'll review it later (anyway I'll have to update the doc and for this to wait for two other tickets. As soon as I get the time I'll check the config stuff, merge it and downgrade the Kea ticket to doc update only so it should get a chance to be merge quickly, BTW very likely before PR#70).

razvan-becheriu commented:
Hi,

I have reverted the changes (spacing and google benchmarks).
Please merge this branch before PR 70

Details form the ticket 5488 (could be useful for documentation):
Modified configure.ac to use pkg-config by default when enabling cassandra backend
(--with-cql)
Added support for statically linking with cassandra cpp driver librabry
--enable-cql-static-lib
Note:
to compile with static cpp-driver, the driver must be compiled with:
cmake .. -DCASS_BUILD_STATIC=ON -DCMAKE_CXX_FLAGS="-fPIC"

If you want to build the packages to use with apt or yum, please add "-DCMAKE_CXX_FLAGS='-fPIC'" to the cmake command by modifying the lines in:

- packaging/debian/rules:
dh_auto_configure -- -DCMAKE_BUILD_TYPE=RELEASE -DCASS_BUILD_STATIC=ON -DCMAKE_CXX_FLAGS='-fPIC' -DCASS_INSTALL_PKG_CONFIG=OFF -DCMAKE_INSTALL_LIBDIR=/usr/lib

- packaging/homebrew/cassandra-cpp-driver.rb:
system "cmake", "-DCMAKE_BUILD_TYPE=RELEASE", "-DCASS_BUILD_STATIC=ON", "-DCMAKE_CXX_FLAGS='-fPIC'", "-DCASS_INSTALL_PKG_CONFIG=OFF", "-DCMAKE_INSTALL_PREFIX:PATH=#{prefix}", "-DCMAKE_INSTALL_LIBDIR=#{lib}", ".."

Regards,
Razvan

fxdupont commented:
Far better but there are 4 points to fix:
 - change the help string from `--with-cql=PATH` to `--with-cql[=PATH]`
 - the `/usr/bin/pkg-config` fixed path does not work for instance on my macOS. I propose to change it by `AC_PATH_PROGS([CQL_CONFIG], [cql_config pkg-config])` followed by a check for the if-not-found case where CQL_CONFIG is empty.
 - the error message should say `cql_config or pkg-config` (BTW you can swap cql_config and pkg-config if you prefer everywhere).
 - if you do not know a system where `-lpthread` is required to build Kea with CQL support (IMHO there is none) please remove it from `tools/cql_config`
Note only the second point needs some work to test: others are cosmetics.

tomaszmrugalski commented on configure.ac:
```diff
@@ -768,10 +781,10 @@ if test "$CQL_CONFIG" != "" ; then
-    CQL_CPPFLAGS="`$CQL_CONFIG --cppflags` -I$CQL_INCLUDEDIR"
-    CQL_LIBS="-L`$CQL_CONFIG --libdir` -lcassandra -luv"
-    CQL_VERSION=`$CQL_CONFIG --version`
+    CQL_INCLUDEDIR=`$CQL_CONFIG --cflags-only-I $cql_lib`
```
Need space between cflags-only and -I

razvan-becheriu replied:
No,
there is a different option for --cflags. 
See --help
This is a separate option:
--cflags-only-I                         output -I flags
The output gives the flags with the -I prefix included.

tomaszmrugalski commented:
Why do we need cql_config script at all? Couldn't configure.ac use pkg-config directly? That way we could get rid of this script.

razvan-becheriu commented:
We need cql_config if we don't want to build and install the cpp-driver.
The 'pkg-config' tool works only with the deb/rpm packages, which must be installed on the system.
The script gives the option to just build using a source directory, without actually installing the packages.
This is also useful when doing static linking with the cpp driver library. In this case, no dependency to an external library is created.

tomaszmrugalski commented:
Understood. Thanks for the explanation. I'll shut up now :)

razvan-becheriu commented:
If you want to completely remove it it is OK with me.
In the DT repo we will keep this option as we really need it.
It is just another option of compiling, just as you already have with
google tests (--with-gtest or --with-gtest-source using installed binaries
or source directory).

It is completely your call if you want to permit compiling from source.

Regards,
Razvan.

On Wed, Mar 14, 2018 at 3:13 PM, Tomek Mrugalski <notifications@github.com>
wrote:

> Understood. Thanks for the explanation. I'll shut up now :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/isc-projects/kea/pull/71#issuecomment-373016091>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ASkW25FawcSrmAroYUj_r15jiXePR3vrks5teReUgaJpZM4ScetQ>
> .
>



-- 
razvan becheriu
[image: qualitance.com]
senior software engineer
+40 (746) 090 415 <+40+(746)+090+415>, qualitance.com
<https://www.qualitance.com/>

tomaszmrugalski commented:
That's fine. Let's keep it. I simply forgot about the use case of compiling from sources. Let's keep the script then.

razvan-becheriu commented:
I'll try to finish it.

fxdupont commented:
Merged in Kea (post 1.4-beta) as trac5488. BTW I pushed this branch to GitHub with the cpp-driver config file for macOS brew, if you have similar package config files please do the same.
